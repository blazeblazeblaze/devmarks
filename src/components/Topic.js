import React from "react";

import Hero from "./Hero";
import TutorialList from "./TutorialList";

const Topic = ({ name, subtitle, items, ...rest }) => (
  <section className="section">
    <div className="container">
      <Hero title={name} subtitle={subtitle} />
    </div>
    <div className="container">
      <TutorialList list={items} />
    </div>
  </section>
);

export default Topic;
