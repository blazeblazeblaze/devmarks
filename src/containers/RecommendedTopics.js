import React, { Component } from "react";
import { connect } from "react-redux";
import TopicList from "../components/TopicList";

import { fetchTopicsIfNeeded, invalidateTopics } from "../actions/topics";
import ErrorMessage from "../components/ErrorMessage";
import DelayedComponent from "../components/DelayedComponent";
import TopicListPlaceholder from "../components/TopicListPlaceholder";

const recommendedTopics = (topics, topic) => {
  const { category, slug } = topic || {
    category: null,
    slug: null
  };

  if (category == null || slug == null) {
    return [];
  }

  return topics
    .filter(topic => topic.category === category)
    .filter(topic => topic.tutorialsCount > 0)
    .filter(topic => topic.slug !== slug)
    .slice(0, 3);
};

const mapStateToProps = ({ topics }, ownProps) => ({
  topics: recommendedTopics(topics.entities, ownProps.topic),
  isLoading: topics.status.isLoading,
  hasErrors: topics.status.hasErrors,
  didInvalidate: topics.status.didInvalidate
});

const RecommendedTopicsPage = class extends Component {
  componentDidMount() {
    this.props.fetchTopicsIfNeeded();
  }

  componentDidUpdate() {
    if (this.props.didInvalidate) {
      this.props.fetchTopicsIfNeeded();
    }
  }

  render() {
    const { isLoading, hasErrors, topics, invalidateTopics } = this.props;

    return hasErrors ? (
      <ErrorMessage onClick={invalidateTopics} />
    ) : (
      <DelayedComponent
        isReady={!isLoading}
        fallbackComponent={<TopicListPlaceholder />}
      >
        <TopicList
          topics={topics}
          fallbackMsg="No matching recommendations found."
        />
      </DelayedComponent>
    );
  }
};

export default connect(mapStateToProps, {
  fetchTopicsIfNeeded,
  invalidateTopics
})(RecommendedTopicsPage);
