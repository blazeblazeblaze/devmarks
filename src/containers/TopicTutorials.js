import React, { Component } from "react";
import { connect } from "react-redux";
import TutorialList from "../components/TutorialList";
import ErrorMessage from "../components/ErrorMessage";
import {
  fetchTutorialsIfNeeded,
  invalidateTutorials
} from "../actions/tutorials";

import DelayedComponent from "../components/DelayedComponent";
import TopicTutorialsPlaceholder from "../components/TopicTutorialsPlaceholder";

const mapStateToProps = ({ tutorials }, { slug }) => {
  const { isLoading, hasErrors, entities } = tutorials[slug] || {
    isLoading: true,
    hasErrors: false,
    entities: []
  };

  return {
    tutorials: entities,
    isLoading,
    hasErrors
  };
};

const TopicTutorials = class extends Component {
  componentDidMount() {
    const { slug } = this.props;

    this.props.fetchTutorialsIfNeeded(slug);
  }

  componentDidUpdate(nextProps) {
    this.props.fetchTutorialsIfNeeded(this.props.slug);
  }

  render() {
    const {
      tutorials,
      hasErrors,
      isLoading,
      invalidateTutorials,
      slug
    } = this.props;

    return hasErrors ? (
      <ErrorMessage
        onClick={e => {
          invalidateTutorials(slug);
        }}
      />
    ) : (
      <DelayedComponent
        isReady={!isLoading}
        fallbackComponent={<TopicTutorialsPlaceholder />}
      >
        <TutorialList tutorials={tutorials} />
      </DelayedComponent>
    );
  }
};

export default connect(mapStateToProps, {
  fetchTutorialsIfNeeded,
  invalidateTutorials
})(TopicTutorials);
