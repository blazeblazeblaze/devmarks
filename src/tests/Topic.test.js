import React from "react";
import Topic from "../components/Topic";
import RecommendedTopics from "../containers/RecommendedTopics";
import TopicTutorials from "../containers/TopicTutorials";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

configure({ adapter: new Adapter() });

describe("<Topic />", () => {
  const slug = "some slug";
  const tree = shallow(<Topic topic={{}} slug={slug} />);

  it("renders Hero", () => {
    expect(tree.find("Hero").props()).toEqual({
      title: "Browse tutorials"
    });
  });

  it("passes slug down to TopicTutorials", () => {
    expect(tree.find(TopicTutorials).props()).toEqual({ slug: slug });
  });

  describe("with given topic", () => {
    it("renders RecommendedTopics", () => {
      expect(tree.find(RecommendedTopics).exists()).toBe(true);
    });

    it("RecommendedTopics receives slug", () => {
      expect(tree.find(RecommendedTopics).props()).toEqual({ topic: {} });
    });
  });

  describe("without topic", () => {
    const tree = shallow(<Topic slug={slug} />);

    it("does not render RecommendedTopics", () => {
      expect(tree.find(RecommendedTopics).exists()).toBe(false);
    });
  });

  it("matches snapshot", () => {
    expect(toJson(tree)).toMatchSnapshot();
  });
});
