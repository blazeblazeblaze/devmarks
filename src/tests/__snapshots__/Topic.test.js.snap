// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Topic /> matches snapshot 1`] = `
<div>
  <Hero
    title="Browse tutorials"
  />
  <Connect(TopicTutorials)
    slug="some slug"
  />
  <Connect(RecommendedTopicsPage)
    topic={Object {}}
  />
</div>
`;
